libutils/threaded_deque.c

```diff
@@ -0,0 +1,641 @@
+    size_t left;                      /**< Current position in deque.      */
+    size_t right;                     /**< Current end of deque.           */
+    size_t size;                      /**< Current size of deque.          */
+    size_t capacity;                  /**< Current memory allocated.       */
+};
+
+static void DestroyRange(ThreadedDeque *deque, size_t start, size_t end);
+static void ExpandIfNecessary(ThreadedDeque *deque);
+
+ThreadedDeque *ThreadedDequeNew(size_t initial_capacity,
+                                void (ItemDestroy) (void *item))
+{
+    ThreadedDeque *deque = xcalloc(1, sizeof(ThreadedDeque));
+
+    if (initial_capacity == 0)
+    {
+        initial_capacity = DEFAULT_CAPACITY;
+    }
+
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+    // enables errorchecking for deadlocks
+    int ret = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to use error-checking mutexes for deque, "
+            "falling back to normal ones (pthread_mutexattr_settype: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_NORMAL);
+    }
+
+    deque->lock = xmalloc(sizeof(pthread_mutex_t));
+    ret = pthread_mutex_init(deque->lock, &attr);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize mutex (pthread_mutex_init: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_destroy(&attr);
+        free(deque->lock);
+        free(deque);
+        return NULL;
+    }
+
+    pthread_mutexattr_destroy(&attr);
+
+    deque->cond_non_empty = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(deque->cond_non_empty, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition (pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(deque->lock);
+        free(deque->cond_non_empty);
+        free(deque);
+        return NULL;
+    }
+
+    deque->cond_empty = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(deque->cond_empty, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition "
+            "(pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(deque->lock);
+        free(deque->cond_empty);
+        free(deque->cond_non_empty);
+        free(deque);
+        return NULL;
+    }
+
+    deque->capacity = initial_capacity;
+    deque->left = 0;
+    deque->right = 0;
+    deque->size = 0;
+    deque->data = xmalloc(sizeof(void *) * initial_capacity);
+    deque->ItemDestroy = ItemDestroy;
+
+    return deque;
+}
+
+void ThreadedDequeDestroy(ThreadedDeque *deque)
+{
+    if (deque != NULL)
+    {
+        ThreadLock(deque->lock);
+        DestroyRange(deque, deque->left, deque->right);
+        ThreadUnlock(deque->lock);
```

Comment:
I agree, and we talked about this in the previous PR as well.

This function should only be called when you are sure all the threads are joined and the program is about to die.

Reply:
Yeah, we discussed this in the `Queue` case. There's no good way to make these things thread-safe without a lock that would be outside of the `Queue/Dequeue` structure which is something we want to avoid.